core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -292,56 +286,44 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
    */
   private class Watchers(val key: Any) {
 
-    private[this] val operations = new LinkedList[T]()
+    private[this] var operations = Vector[T]()
 
-    def watched: Int = operations synchronized operations.size
+    def watched: Int = synchronized(operations.size)
 
     // add the element to watch
     def watch(t: T) {
-      operations synchronized operations.add(t)
+      synchronized {
+        operations = operations :+ t
+      }
     }
 
     // traverse the list and try to complete some watched elements
     def tryCompleteWatched(): Int = {
-
-      var completed = 0
-      operations synchronized {
-        val iter = operations.iterator()
-        while (iter.hasNext) {
-          val curr = iter.next()
-          if (curr.isCompleted) {
-            // another thread has completed this operation, just remove it
-            iter.remove()
-          } else if (curr synchronized curr.tryComplete()) {
-            completed += 1
-            iter.remove()
-          }
-        }
+      // call tryComplete without holding the lock to avoid potential deadlocks
+      var completedAlready = 0
+      var completedNow = 0
+      for (op <- synchronized(operations)) {
```

Comment:
@becketqin, there is no full copy on each `watch`. Persistent collections like `Vector` do a lot of internal sharing and `append` in particular is `log(n)` for `Vector`. We are doing a full copy on `filterNot`, but that only happens when some operations have completed and should be removed from the collection.

A benefit of the current implementation is that we only hold locks very briefly if no operation has been completed.

If the performance results look good, this seems like a reasonable implementation. Using mutable collections while keeping it simple would probably involve copying the collection before iterating, which seems worse.